src/modules/musig/main_impl.h

```diff
@@ -12,24 +12,15 @@
 #include "include/secp256k1_musig.h"
 #include "hash.h"
 
-int secp256k1_musig_sort_pubkeys(const secp256k1_context* ctx, secp256k1_xonly_pubkey *pubkeys, size_t n_pubkeys) {
-    size_t i, j;
-    secp256k1_xonly_pubkey key;
+static int secp256k1_musig_sort_pubkeys_cmp (const void *p1, const void *p2) {
+  return secp256k1_memcmp_var(p1, p2, sizeof(secp256k1_xonly_pubkey));
```

Comment:
from `secp256k1_xonly_pubkey`:

    *  The exact representation of data inside is implementation defined and not
    *  guaranteed to be portable between different platforms or versions. It is
    *  however guaranteed to be 64 bytes in size, and can be safely copied/moved.
    *  If you need to convert to a format suitable for storage, transmission, or
    *  comparison, use secp256k1_xonly_pubkey_serialize and
    *  secp256k1_xonly_pubkey_parse.

Not only is sorting this way an API violation, I believe that you will literally get different orders on architectures of different endianness.

Reply:
Indeed, and the tests can be improved then. Maybe we should sync this with https://github.com/bitcoin-core/secp256k1/pull/850

Reply:
Just be aware that `secp256k1_xonly_pubkey_cmp` cannot be used with `qsort` due to its (nearly useless) `ctx` argument.  I'm not sure if this comment belongs here or there.

Reply:
I suppose here it could be wrapped up with a `secp256k1_context_no_precomp` context for the `qsort` call.

Reply:
Ugh, that's what you get when you use memcmp for equality testing too often. CI caught it.

All functions operating on pubkeys need a context, because its required for `pubkey_load`. If I understand the `no_precomp` suggestion correctly, it has the downside that if `pubkey_load` fails, the illegal callback of that temporary context will be called instead of whatever the caller of `sort_pubkeys` may have set. So I see the two options

1. Call `pubkey_load` for all pubkeys with the correct context before qsorting to make sure that it won't fail when called in the comparator.
2. Revert back to insertion sort (temporary). My preferred solution because it doesn't make fragile assumptions about the behavior of the contexts.

Reply:
I figure you would do 1.

Reply:
1b. Define a `secp256k1_pubkey_load_no_callback` that simply returns 0 but doesn't need a context to access a callback.

These dynamic callbacks are annoying in a few other places, e.g. you can't call for them for illegal arguments when *creating* a context (chicken-and-egg problem). I'll probably make a suggestion to get rid of the callbacks in the contexts and only have compile-time overrides for the default callbacks. I don't think anyone needs the ability to change the callbacks at runtime.

Reply:
By the way, once keys are loaded, we  can just use `secp256k1_fe_cmp_var`.

Reply:
@roconnor-blockstream why?

Reply:
@real-or-random (a) `secp256k1_pubkey_load_no_callback` would mean not using `secp256k1_xonly_pubkey_cmp` or involve changing the definition in that PR (though these options are doable I guess).

(b) There is no place to load all the pubkeys into in order to sort by `secp256k1_fe_cmp_var`.

@jonasnick Because you get you use `qsort`[1] instead of insertion sort, and I'm dubious of your claim that it would make fragile assumptions of the behavior of contexts.  Maybe I don't understand what you mean by "fragile assumptions", but if `secp256k1_xonly_pubkey_cmp` cannot compare xonly pubkeys that can all be successfully loaded, then something is wrong with the specification of `secp256k1_xonly_pubkey_cmp`.

But hey, if you want to use insertion sort, I'm not going to get in your way.

[1] not that qsort is that great to begin with: https://travisdowns.github.io/blog/2019/05/22/sorting.html

Reply:
Oh i see, your concern is that you want to be open to the possibility that loading a pubkey requires a precomputation table or something.

    /** A simple secp256k1 context object with no precomputed tables. These are useful for
     *  type serialization/parsing functions which require a context object to maintain
     *  API consistency, but currently do not require expensive precomputations or dynamic
     *  allocations.
     */
    SECP256K1_API extern const secp256k1_context *secp256k1_context_no_precomp; 

Loading an x-only pubkey seems like it qualifies as "type serialization/parsing functions which require a context object to maintain API consistency, but currently do not require expensive precomputations or dynamic allocations."  If you consider this use as fragile then you might as well open a PR to eliminate `secp256k1_context_no_precomp` because I cannot imagine a safer place to use it than here.

Reply:
But as mentioned above, just using `secp256k1_context_no_precomp` isn't safe at all in this case. When encountering an illegal pubkey, the default callback aborts the program instead of calling the user's callback. The assumption inside `pubkey_load` is that the context is the user's not some temporary context. We can fix this for now by trying to load all pubkeys with the user's context object before calling qsort just to ensure that there's no illegal pubkey and the default illegal callback isn't called on the temporary context.

The main issue is that qsort doesn't have a powerful enough interface, so I don't see why we'd want to hack around this instead of just using a function with the right interface.

Reply:
@jonas as you yourself mention above you can "Call pubkey_load for all pubkeys with the correct context before qsorting to make sure that it won't fail when called in the comparator." Subsequent calls to pubkey_load can then use `secp256k1_context_no_precomp` safely.

Anyhow if you think "[deliberately quadratic](https://accidentallyquadratic.tumblr.com/)] is okay for this code, then I won't stop you.

Reply:
> (b) There is no place to load all the pubkeys into in order to sort by `secp256k1_fe_cmp_var`.

Indeed but I didn't want to suggest this. My suggestion was to always perform two loads inside `secp256k1_xonly_pubkey_cmp` and then to call `secp256k1_fe_cmp_var`. That's not great for performance but at least it's the same complexity as `qsort`.

Reply:
Yes, this hack works in this specific case, but it still breaks the assumption that pubkey_load is called on the users context object. The quadratic sort function was certainly only meant to be temporary for the purpose of implementing the spec. I can see the point that until it is replaced this can be viewed as a DoS vector. Do you know of a drop-in sorting implementation that we could use?

Reply:
@real-or-random How do you "perform a load" inside `secp256k1_xonly_pubkey_cmp` without a context object?

Reply:
I don't see what assumption break.  the preprocessing loop *does* call pubkey_load on the user's context object.  The only "assumption" I'm making is that `secp256k1_xonly_pubkey_cmp` doesn't call anything in the context object when the two keys being compared are loadable.

Reply:
Finding an off the shelf quasi-linear, in-place, no-allocation, "reentrant" sorting function with a suitable license might be a challenge.  glibc's quicksort has a GPL license.  FreeBSD's `qsort_r` [might not be quasi-linear](http://calmerthanyouare.org/2014/06/11/algorithmic-complexity-attacks-and-libc-qsort.html).  Though presumably it is fixed by now.

Reply:
My specific worry was that the context could be used to store additional things, like the number of `pubkey_load`s which would be wrong after `pubkey_sort`. Seeing that the context is const almost everywhere this is likely a non-issue.

> Do you know of a drop-in sorting implementation that we could use?

If the answer is no, I can implement the qsort hack for now with a TODO comment to fix this at some point so we have at least that quadratic DoS issue fixed.

Reply:
> @real-or-random How do you "perform a load" inside `secp256k1_xonly_pubkey_cmp` without a context object?

Using `secp256k1_pubkey_load_no_callback`. Sorry I should be verbose.

Reply:
For the record, I floated an idea on IRC that `secp256k1_xonly_pubkey_cmp` and friends be amended to explicitly allow a `NULL` context to be passed, and won't invoke the callback upon an error.

Reply:
That's not unreasonable, but it's not clear to me how it helps here. If we'd call `secp256k1_xonly_pubkey_cmp` with ctx=NULL, we'd still want to separately try loading with the user's context. Otherwise, there are API inconsistencies where an illegal pubkey provided to `pubkey_serialize` invokes the callback, but when provided to `pubkey_sort` the function just returns 0.

Reply:
I suppose that is a question for @real-or-random, but what is the API inconsistency exactly?

Assuming we went with this modified real-or-random-1b plan, we simply observe that  pubkey_serialize zeros the output buffer when it returns 0.  Thus all invalid keys would compare as less than every valid pubkey, and all invalid keys would compare as equal, and get sorted in an implementation defined order.  This behavour can be clearly specified in `secp256k1_xonly_pubkey_cmp` and `pubkey_sort`.  Indeed this is already what happens if/when the `illegal_callback` returns.  After sorting with erroneous keys, if you try to musig assemble the pubkeys, you will fail then anyways.

But if you have erroneous blobs of memory in your xonly-pubkey array, then every proposal so far dictates garbage-in garbage-out.  As it stands already nothing in `secp256k1_xonly_pubkey_load` checks that the memory blob is on curve or has an even y-coordinate.  The `secp256k1_xonly_pubkey_load` function only fails when the blob of memory has a 0 x-coordinate, which is a scenario that should only happen if someone has failed to check the return code of some operation.  In which case I don't see what's wrong with just specifying that all these invalid keys get sorted to the beginning of the array.  They will still be just as invalid when they are used later.  There seems to be no harm in sorting them.

Reply:
If we do choose to go wtihout the preprocessing step of trying to load every key before sorting, we would simply want to remove the ctx argument to make it clear to everyone that no callback is going to be invoked.

Reply:
The inconsistency would be as follows
```
secp256k1_xonly_pubkey pk;
memset(pk, 0, sizeof(pk));
secp256k1_xonly_serialize(ctx, buf, pk); /* operates on illegal pubkey, calls illegal callback, returns 0 */
secp256k1_pubkey_sort(ctx, buf, pk); /* operates on illegal pubkey, does not call illegal callback, doesn't return 0 */
```

It seems that you're saying that this wouldn't be a problem because the callback would be invoked after pubkey_sort, in key aggregation or serialization. 

> remove the ctx argument 

Remove it where?

Reply:
Yes, I'm saying that sorting illegal pubkeys could be considered okay because the only thing you can do with a sorted array with illegal pubkeys is to do something else with them and whatever that is will trigger the callback.

The modified-1b proposal would end up looking something like

    static int secp256k1_musig_sort_pubkeys_cmp (const void *p1, const void *p2) {
        return secp256k1_xonly_pubkey_cmp(NULL, p1, p2);
    }
    
    void secp256k1_musig_sort_pubkeys(secp256k1_xonly_pubkey *pubkeys, size_t n_pubkeys) {
        VERIFY_CHECK(pubkeys != NULL);

        qsort(pubkeys, n_pubkeys, sizeof(*pubkeys), secp256k1_musig_sort_pubkeys_cmp);
    }

Reply:
Possible problem:  C doesn't make any complexity guarantees about `qsort`'s complexity.  In particular [glibc's `_quicksort`](https://elixir.bootlin.com/glibc/latest/source/stdlib/qsort.c#L74), which `qsort` falls back to when malloc fails, doesn't implement the median of medians pivot which is usually required to bring down the worst case complexity to quasi-linear.

Would you like me to write you an in-place heap sort?

Reply:
@roconnor-blockstream good catch. Yes that would be great. Standard in-place and iterative heap sort should be easy to review.